### PR TITLE
[libc++][numeric] P0543R3: Saturation arithmetic

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements: https://wg21.link/P0543R3
- https://eel.is/c++draft/numeric.sat

Additional references:
- Division: https://eel.is/c++draft/expr.mul#4
- Arithmetic conversions: https://eel.is/c++draft/expr.arith.conv#1
- Clang builtins: https://clang.llvm.org/docs/LanguageExtensions.html#builtin-functions

Depends on: https://github.com/llvm/llvm-project/pull/78086

### CODE REVIEW

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (frederick-vs-ja) [inline libcxx/include/__numeric/saturation_arithmetic.h:28]: I think we should use `remove_cv_t` here. The current standard wording says
> _Constraints_: `T` is a signed or unsigned integer type ([basic.fundamental]).

which clearly rejects reference types.

(Also the name seems misleading as it's intended to accept extended integer types.)

----
BTW, after looking into `<__utility/cmp.h>`, it seems to me that the newly added `__libcpp_standard_integer` should be essentially same as `__is_safe_integral_cmp`.

https://github.com/llvm/llvm-project/blob/7dd4d28e4196fad83ed78ea342d65e7eaec4a6f1/libcxx/include/__utility/cmp.h#L35-L51

Perhaps we can only introduce one concept.

Comment (frederick-vs-ja) [reply libcxx/include/__numeric/saturation_arithmetic.h:28]: Oh, the core wording in [[basic.fundamental]](https://eel.is/c++draft/basic.fundamental) looks weird... It seems that currently `const int` is neither a signed integer type nor a standard integer type because the terms only refer to cv-unqualified types.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (mordante) [inline libcxx/include/__numeric/saturation_arithmetic.h:107]: pedantic: this handles the non-overflowing `__x == std::numeric_limits<_Rp>::min()` too.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.pass.cpp:41]: I would like to see a few more edge cases, -1, 0 , and 1. We should test with the maximum for both the `__x_` and the `__y`. I also like to see cases that overflow when 2 large numbers are added. For example 
__x = max() / 2 + 10
__x = max() / 2 + 5

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.verify.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.verify.cpp:35]: Can you test for the following types too? `char8_t`, `float`, `double`, and `long double`.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.assert.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.assert.pass.cpp:13]: I don't see a test for
```
Preconditions: y != 0 is true. 
...
 A function call expression that violates the precondition in the Preconditions: element is not a core constant expression (7.7 [expr.const]). 
```
did I miss it?

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (mordante) [inline libcxx/include/__numeric/saturation_arithmetic.h:105]: We can avoid run-time comparisons when saturation is impossible.
```suggestion
_LIBCPP_HIDE_FROM_ABI constexpr _Rp saturate_cast(_Tp __x) noexcept {
  // Saturation is impossible.
  if constexpr (std::cmp_greater_equal(std::numeric_limits<_Rp>::min(), std::numeric_limits<_Tp>::min()) && 
                         std::cmp_less_equal(std::numeric_limits<_Rp>::max(), std::numeric_limits<_Tp>::max()))
    return static_cast<_Rp>(__x);
  else {
    ....
```

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for working on this. I'm happy to accept the patch without 128-bit integral support and add that later.

I have not looked closely at the tests, I'd like a bit more test cases before looking at all of them.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (philnik777) [reply libcxx/include/__numeric/saturation_arithmetic.h:32]: Couldn't we just extend `__builtin_add_overflow` to accept extended integer types?

Comment (mordante) [reply libcxx/include/__numeric/saturation_arithmetic.h:32]: How does that differ from what I asked Aaron?

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (mordante) [reply libcxx/include/__numeric/saturation_arithmetic.h:105]: If it can be shown Clang optimizes this I'm happy with it.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (H-G-Hristov) [reply libcxx/include/__numeric/saturation_arithmetic.h]: Pre-requisite added here: https://github.com/llvm/llvm-project/pull/78086
I'll update this patch after the above is merged.

**Comment by H-G-Hristov** [comment]
Thank you for the review!

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (H-G-Hristov) [reply libcxx/include/__numeric/saturation_arithmetic.h:27]: > These builtins don't work with 128-bit integrals. @AaronBallman are you happy to receive patches for this in Clang? If so what is the proper name for `__builtin_uaddXX_overflow` and `____builtin_saddXX_overflow`? Obviously there should be patches for `sub` and `mul`.
> 
> @H-G-Hristov if acceptable for Clang do you want to do the work?

Yes, I could try to do the work. Please note I have no experience with Clang yet.

Comment (philnik777) [reply libcxx/include/__numeric/saturation_arithmetic.h:28]: @mordante It sounded to me like you've asked for additional builtins for extended integer types instead of extending the existing one.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (philnik777) [reply libcxx/include/__numeric/saturation_arithmetic.h:101]: Isn't that backwards from what we usually do? Unless it is shown that things improve the performance, we don't add optimizations. Also, FWIW: https://godbolt.org/z/7rasjdP7j

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp
Comment (philnik777) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp:81]: We don't want to test the compiler in our tests. There is no need to check that division by zero results in an expression that's not constant AFAICT. The check for the `constexpr` expression can simply be tested by putting it into a scope that is forced to be constant evaluated.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (mordante) [reply libcxx/include/__numeric/saturation_arithmetic.h:28]: I asked both, extend `__builtin_add_overflow` for 128 bit, but typically there is a set like
`__builtin_[su]addl{0,2}_overflow` too, I think it would be good to keep that pattern from Clang's PoV. Having a new builtin also makes it easy for us to query whether `__builtin_add_overflow` has 128-bit support.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (mordante) [reply libcxx/include/__numeric/saturation_arithmetic.h:101]: I don't think it's backwards, I'm quite sure we have used `if constexpr` to do optimizations. I see it only matters on `-O0`. Maybe add a comment the compiler can do this optimization.

Comment (philnik777) [reply libcxx/include/__numeric/saturation_arithmetic.h:101]: We've used `if constexpr` for optimizations, but generally you have to show that an optimization is actually an improvement. You basically requested that it's shown that your change isn't one.

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (AaronBallman) [reply libcxx/include/__numeric/saturation_arithmetic.h:28]: `__builtin_add_overflow` already works with 128-bit integers: https://godbolt.org/z/non8eMTfj, as well as bit-precise integers: https://godbolt.org/z/dxs7h4G5b so I'm not certain there are Clang changes needed?

Comment (AaronBallman) [reply libcxx/include/__numeric/saturation_arithmetic.h:28]: You'll probably want to add test coverage for `_BitInt` as well given that it's an extension Clang supports in C++.

Comment (mordante) [reply libcxx/include/__numeric/saturation_arithmetic.h:28]: > `__builtin_add_overflow` already works with 128-bit integers: https://godbolt.org/z/non8eMTfj, as well as bit-precise integers: https://godbolt.org/z/dxs7h4G5b so I'm not certain there are Clang changes needed?

Thanks @AaronBallman there are indeed no changes needed. I just need to learn how to read ;-)

File: libcxx/include/__numeric/saturation_arithmetic.h
Comment (H-G-Hristov) [reply libcxx/include/__numeric/saturation_arithmetic.h:101]: OK. I'll remove the optimization for now. On optimized builds the difference according to my quick, stupid benchmark is near zero and on unoptimized builds less than 10%.
https://quick-bench.com/q/QFdMFKaw_5OnmeDqB_BMkh0BaSg

Please comment if you disagree.

**Comment by H-G-Hristov** [comment]
How do I fix the following generated files check error:

```diff
[8/9] cd /home/runner/work/llvm-project/llvm-project/build/check-generated-output/libcxx/utils && /usr/bin/python3.10 /home/runner/work/llvm-project/llvm-project/libcxx/utils/generate_iwyu_mapping.py
[9/9] cd /home/runner/work/llvm-project/llvm-project/build/check-generated-output/libcxx/utils && /usr/bin/python3.10 /home/runner/work/llvm-project/llvm-project/libcxx/utils/generate_escaped_output_table.py /home/runner/work/llvm-project/llvm-project/libcxx/include/__format/escaped_output_table.h
diff --git a/libcxx/include/libcxx.imp b/libcxx/include/libcxx.imp
index 8616f9639..45fa4a954 100644
--- a/libcxx/include/libcxx.imp
+++ b/libcxx/include/libcxx.imp
@@ -559,6 +559,7 @@
   { include: [ "<__numeric/pstl_reduce.h>", "private", "<numeric>", "public" ] },
   { include: [ "<__numeric/pstl_transform_reduce.h>", "private", "<numeric>", "public" ] },
   { include: [ "<__numeric/reduce.h>", "private", "<numeric>", "public" ] },
+  { include: [ "<__numeric/saturation_arithmetic.h>", "private", "<numeric>", "public" ] },
   { include: [ "<__numeric/transform_exclusive_scan.h>", "private", "<numeric>", "public" ] },
   { include: [ "<__numeric/transform_inclusive_scan.h>", "private", "<numeric>", "public" ] },
   { include: [ "<__numeric/transform_reduce.h>", "private", "<numeric>", "public" ] },
```

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.pass.cpp:26]:
```suggestion
  static_assert(noexcept(std::add_sat(minVal, maxVal)));
```
please check all noexcept test for the copy paste issue.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/add_sat.pass.cpp:30]: How about testing the result type once and then directly assert the result. Something like
```
  std::same_as<IntegerT> decltype(auto) sum = std::add_sat(IntegerT{0}, IntegerT{0});
  assert(sum == IntegerT{0});

  assert(std::add_sat(IntegerT{1}, IntegerT{0}) == IntegerT{1});
  assert(std::add_sat(IntegerT{1}, IntegerT{1}) == IntegerT{0});
  assert(std::add_sat(IntegerT{1}, IntegerT{0}) == IntegerT{-1});
  assert(std::add_sat(IntegerT{1}, IntegerT{-1}) == IntegerT{0});
  assert(std::add_sat(IntegerT{1}, IntegerT{1}) == IntegerT{1});
  ...
```
Currently every test entry is 5 lines, which makes the test quite long. (I'm not against long tests in general, but here the test density is a bit low.)

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp:52]: why not `-maxVal`?

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp:109]: I see quite a bit of copy paste with `sum`. That should be `quot`, or not use a temporary value.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/div_sat.pass.cpp:234]:
```suggestion
  // Unsigned integer division never overflows
```

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.compile.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.compile.pass.cpp:26]:
```suggestion
  static_assert(CanDo<R, T>);
  static_assert(CanDo<T, R>);
```

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp:21]: I'm not fond of this test. IMO it's a bit too smart and it makes it hard to understand what is tested.
What I often do for these kind of tests is something like
```
assert(std::saturate_cast<signed char, int>(-255) == -128);
assert(std::saturate_cast<signed char, int>(-128) == -128);
assert(std::saturate_cast<signed char, int>(0) == 0);
assert(std::saturate_cast<signed char, int>(127) == 127);
assert(std::saturate_cast<signed char, int>(255) == 127);
```

and similar for other conversions. I think that is easier to follow. (Note I like the add, sub, mul, and div test.)

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks I think this is getting close, I would like to see it again after addressing the comments.

**Comment by H-G-Hristov** [comment]
I did some re-imagining of the tests, there is a some redundancy but it is cleared that nothing was missed.
I also used the Clang 18's "Placeholder variables with no name" to make the test a bit cleaner, so I disabled the unsupported compilers.
The tests are formatted in tabular form manually to make them easier to read.

If the above is unacceptable, it can be fixed easily. For now it makes it easier to read and review.

**Review by mordante — CHANGES_REQUESTED** [review]
> I did some re-imagining of the tests, there is a some redundancy but it is cleared that nothing was missed.  The tests are formatted in tabular form manually to make them easier to read.

Redundant code in tests when if improves readability is not too bad. I really like how the new testing code looks. It was a lot easier to follow. Thanks a lot!

>  I also used the Clang 18's "Placeholder variables with no name" to make the test a bit cleaner, so I disabled the unsupported compilers.
> If the above is unacceptable, it can be fixed easily. For now it makes it easier to read and review.

IMO we should not disable tests for compilers, only for the reason to be able to use newer language features in the test. Note if the header requires language features I'm fine to disable that for a compiler. In this case we can easily allow clang-17 with a minor change as pointed out in the review.

The patch is getting close to be ready.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp
Comment (H-G-Hristov) [reply libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp]: These are needed specifically as these values can vary but I used the C integer constants wherever applicable.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp
Comment (H-G-Hristov) [reply libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp]: I want to test against the specific type, so I added constants for the missing literals.

**Comment by H-G-Hristov** [comment]
There is a problem with the test. Working on it.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp
Comment (mordante) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.sat/saturate_cast.pass.cpp:76]:
```suggestion
  // here and below...
```

**Review by mordante — APPROVED** [review]
LGTM modulo one typo.

**Comment by vvereschaka** [comment]
@Zingam ,
these changes break the aarch64 cross builder with the failed test `llvm-libc++-static.cfg.in::mul_sat.pass.cpp`
* https://lab.llvm.org/buildbot/#/builders/119/builds/16778

```
# .---command stderr------------
# | ld.lld: error: undefined symbol: __muloti4
# | >>> referenced by mul_sat.pass.cpp
# | >>>               C:\Users\buildbot\AppData\Local\Temp\lit-tmp-vspdi658\mul_sat-857336.o:(__int128 std::__2::mul_sat[abi:ne180000]<__int128>(__int128, __int128))
# | clang++: error: linker command failed with exit code 1 (use -v to see invocation)
# `-----------------------------
```

**Comment by vvereschaka** [comment]
@H-G-Hristov, @Zingam any updates?

**Comment by vvereschaka** [comment]
Looks like I see the problem. The test configuration must be updated on the builder.
Sorry for the false alarm.

**Pull request merged**
